### PR TITLE
add YandexMapKit support

### PR DESCRIPTION
**Pull request by petropavel13**

### CODE REVIEW

**Review by Viveron — APPROVED** [review]

File: ClusterKit.podspec [hunk 1]
```diff
@@ -35,19 +35,27 @@ Pod::Spec.new do |s|
...
 
   s.subspec 'Mapbox' do |ss|
     ss.platform = :ios, '9.0'
     ss.dependency 'ClusterKit/Core'
-    ss.dependency 'Mapbox-iOS-SDK', '~> 4.6'
+    ss.dependency 'Mapbox-iOS-SDK', '~> 4.7'
     ss.source_files = 'ClusterKit/Mapbox'
   end
 
+  s.subspec 'YandexMapKit' do |ss|
```
Comment (maxep) [inline ClusterKit.podspec]: Like `GoogleMaps` sdk (https://github.com/googlemaps/google-maps-ios-utils/issues/23) `YandexMapKit` is statically built, which mean we can't use them as subspec dependency yet. Better to keep both `GoogleMaps` and `YandexMapKit` commented until both of them are dynamically built!

File: ClusterKit.podspec [hunk 2]
```diff
@@ -35,19 +35,27 @@ Pod::Spec.new do |s|
...
+   ss.dependency 'ClusterKit/Core'
+   ss.dependency 'GoogleMaps', '~> 2.7'
+   ss.source_files = 'ClusterKit/GoogleMaps'
+  end
 
   s.subspec 'Mapbox' do |ss|
     ss.platform = :ios, '9.0'
     ss.dependency 'ClusterKit/Core'
-    ss.dependency 'Mapbox-iOS-SDK', '~> 4.6'
+    ss.dependency 'Mapbox-iOS-SDK', '~> 4.7'
```
Comment (maxep) [inline ClusterKit.podspec]: Better keep that for another PR.

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+- (void)removeClusters:(NSArray<CKCluster *> *)clusters {
+    for (CKCluster *cluster in clusters) {
+        [self removeCluster:cluster];
+    }
+}
+
+- (void)selectCluster:(CKCluster *)cluster animated:(BOOL)animated {
+    YMKPlacemarkMapObject *placemark = [self.placemarks objectForKey:cluster];
+    if (placemark) {
+        [self.delegate selectPlacemark:placemark
```
Comment (maxep) [inline ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m]: Should check if delegate respond to selector.

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+                            forCluster:cluster
+                              animated:animated];
+    }
+}
+
+- (void)deselectCluster:(CKCluster *)cluster animated:(BOOL)animated {
+    YMKPlacemarkMapObject *placemark = [self placemarkForCluster:cluster];
+    if (placemark) {
+        [self.delegate deselectPlacemark:placemark
```
Comment (maxep) [inline ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m]: Should check if delegate respond to selector.

File: Examples/Example-swift/YandexMapViewController.swift [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+                                 cameraUpdateSource: YMKCameraUpdateSource,
+                                 finished: Bool) {
+
+        guard isMapLoaded, finished else {
+            return
+        }
+
+        mapView.clusterManager.updateClustersIfNeeded()
+
+        // workaround (https://github.com/yandex/mapkit-ios-demo/issues/23)
```
Comment (maxep) [inline Examples/Example-swift/YandexMapViewController.swift:98]: Good to know!

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            completionBlock(finished);
+        };
+    }
+
+    if ([self.clusterManager.delegate respondsToSelector:@selector(clusterManager:performAnimations:completion:)]) {
+        [self.clusterManager.delegate clusterManager:self.clusterManager
+                                   performAnimations:animationsBlock
+                                          completion:completionBlock];
+    } else {
+        // actually, it didn't work (https://github.com/yandex/mapkit-ios-demo/issues/21)
```
Comment (maxep) [inline ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m:253]: So animations are not yet supported on `YandexMapKit`?

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.h [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+ @param mapView A map view object requesting the marker.
+ @param cluster The cluster to represent.
+
+ @return An object inheriting from GMSMarker that the map view can use for the specified cluster.
+ */
+- (__kindof YMKPlacemarkMapObject *)mapView:(YMKMapView *)mapView placemarkForCluster:(CKCluster *)cluster;
+
+@end
+
+@protocol YMKMapViewDelegate <NSObject>
```
Comment (maxep) [inline ClusterKit/YandexMapKit/YMKMapView+ClusterKit.h]: I don't really see the point of having `YMKMapViewDelegate` to manage selection/deselection as this should be handled by the `YMKMapObjectTapListener`. Is there a particular use case that you are thinking of?

**Review by maxep — CHANGES_REQUESTED** [review]
Thank you very much @petropavel13 ! this is an awesome contribution and it is very well implemented!
I've some comments following my review.

File: ClusterKit.podspec (hunk 1 above)
Comment (petropavel13) [reply ClusterKit.podspec]: got it. thanks!
will add your comment to podspec to clarify the problem for everyone else

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m (hunk 3 above)
Comment (petropavel13) [reply ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m]: Can you explain why?

There is a protocol with non-optional requirements:
```objc
@protocol YMKMapViewDelegate <NSObject>

- (void)selectPlacemark:(YMKPlacemarkMapObject *)placemark
             forCluster:(CKCluster *)cluster
               animated:(BOOL)animated;

- (void)deselectPlacemark:(YMKPlacemarkMapObject *)placemark
               forCluster:(CKCluster *)cluster
                 animated:(BOOL)animated;

@end
```

and delegate object that conforms it:
```objc
@property(nonatomic, weak) IBOutlet id<YMKMapViewDelegate> delegate;
```

it gives us, at least, compile-time safety, isn't it?

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m (hunk 6 above)
Comment (petropavel13) [reply ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m:253]: apparently yes :/
or I don't know how to make it work

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.h (hunk 7 above)
Comment (petropavel13) [reply ClusterKit/YandexMapKit/YMKMapView+ClusterKit.h]: Agree. No. Maybe there will be a particular case, so I implemented that in advance.
I'll remove this delegate.

**Comment by petropavel13** [comment]
Thanks for quick answer and code review!

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m (hunk 3 above)
Comment (maxep) [reply ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m]: Yes you right, my bad!

File: ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m (hunk 6 above)
Comment (maxep) [reply ClusterKit/YandexMapKit/YMKMapView+ClusterKit.m:253]: Hopefully, that's gonna be available in a near future!

**Review by maxep — APPROVED** [review]
Thanks for the quick update, this all looks good to me! thanks again for that contribution

**Pull request merged**